fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -1072,14 +1082,15 @@ ACTOR Future<bool> checkExclusion(Database db,
 		return false;
 	}
 
-	if (ssExcludedCount == ssTotalCount ||
-	    (1 - worstFreeSpaceRatio) * ssTotalCount / (ssTotalCount - ssExcludedCount) > 0.9) {
+	double finalFreeRatio = 1 - (totalKvStoreUsedBytes / (totalKvStoreUsedBytesNonExcluded + totalKvStoreFreeBytes));
```

Comment:
Did you weigh the idea of including available bytes here? This would be where we change `totalKvStoreUsedBytes` -> `totalKvStoreUsedBytes + totalKvStoreFreeBytes - totalKvStoreAvailableBytes`. This will make the free space ratio larger if there are available bytes within the storage files, so it's even less pessimistic. I think it should be fairly safe to do, and it would be more accurate. We could also omit this change and be more conservative.

@sfc-gh-satherton Does redwood need us to account for available space in this check?

Reply:
I'm not sure how well FDB can utilize the fragmented space inside used files, so just went with bit conservative calculation. I believe it should be fairly safe though, as its probably looking for free pages inside storage files first and they are always up-to-date?

Reply:
At least for sqlite, available space is readily reusable and will be consumed before new pages are added. For the processes being removed, their internal space marked available shouldn't result in any extra consumption on the remaining processes once the data is moved.

I suspect the same is true for Redwood, but @sfc-gh-satherton could answer better. Additionally, this may be needed for Redwood if it's never returning its available space and shrinking its files.

Reply:
Redwood doesn't shrink files, so internal free space must be considered, however Redwood actually deducts the internal free space from the file size in populating `StorageBytes::used`.  This is inconsistent with what SQLite does, unfortunately.   However, they both are consistent for `StorageBytes::available` space which in both cases includes filesystem free space + internal free space ready for immediate re-use.

Reply:
is `StorageBytes::used` is equivalent to `StorageBytes::stored` then for Redwood?

Reply:
I see, these probably shouldn't be disagreeing on the definitions of these fields given that people can use them to do math like this.

It seems like the only thing we can do to make this consistent is to change the definition of the redwood field or introduce some new fields that capture this data and are consistent between the two.

Reply:
@vishesh I'll let you decide how to proceed in light of this. If you want better accuracy for SQLite (and maybe the memory storage engine), then I think this should account for available space. Redwood can fix the value published for used bytes accordingly.

If you'd rather keep it as currently written to be more conservative, then that's fine too. Redwood will work ok in that scheme (it should still probably update used, though) and the other storage engines may overestimate the amount of space consumed, but I think it would still be better than the existing logic.

Reply:
I think we can merge this as of now as I ran tests on our kube tests. I will create a issue to update with new one and check against other storage engines, as probably should check against Rocks too.